proxy_worker/dispatcher.py

```diff
@@ -0,0 +1,521 @@
+        current_task = asyncio.current_task(loop)
+        if current_task is not None:
+            task_invocation_id = getattr(current_task,
+                                         ContextEnabledTask.AZURE_INVOCATION_ID,
+                                         None)
+            if task_invocation_id is not None:
+                return task_invocation_id
+
+    return getattr(_invocation_id_local, 'invocation_id', None)
+
+
+class AsyncLoggingHandler(logging.Handler):
+    def emit(self, record: logging.LogRecord) -> None:
+        # Since we disable console log after gRPC channel is initiated,
+        # we should redirect all the messages into dispatcher.
+
+        # When dispatcher receives an exception, it should switch back
+        # to console logging. However, it is possible that
+        # __current_dispatcher__ is set to None as there are still messages
+        # buffered in this handler, not calling the emit yet.
+        msg = self.format(record)
+        try:
+            Dispatcher.current.on_logging(record, msg)
+        except RuntimeError as runtime_error:
+            # This will cause 'Dispatcher not found' failure.
+            # Logging such of an issue will cause infinite loop of gRPC logging
+            # To mitigate, we should suppress the 2nd level error logging here
+            # and use print function to report exception instead.
+            print(f'{CONSOLE_LOG_PREFIX} ERROR: {str(runtime_error)}',
+                  file=sys.stderr, flush=True)
+
+
+@dataclass
+class WorkerRequest:
+    name: str
+    request: str
+    properties: Optional[dict[str, typing.Any]] = None
+
+
+class DispatcherMeta(type):
+    __current_dispatcher__: Optional["Dispatcher"] = None
+
+    @property
+    def current(cls):
+        disp = cls.__current_dispatcher__
+        if disp is None:
+            raise RuntimeError('no currently running Dispatcher is found')
+        return disp
+
+
+class Dispatcher(metaclass=DispatcherMeta):
+    _GRPC_STOP_RESPONSE = object()
+
+    def __init__(self, loop: AbstractEventLoop, host: str, port: int,
+                 worker_id: str, request_id: str,
+                 grpc_connect_timeout: float,
+                 grpc_max_msg_len: int = -1) -> None:
+        self._loop = loop
+        self._host = host
+        self._port = port
+        self._request_id = request_id
+        self._worker_id = worker_id
+        self._grpc_connect_timeout: float = grpc_connect_timeout
+        self._grpc_max_msg_len: int = grpc_max_msg_len
+        self._old_task_factory: Optional[Any] = None
+
+        self._grpc_resp_queue: queue.Queue = queue.Queue()
+        self._grpc_connected_fut = loop.create_future()
+        self._grpc_thread: Optional[threading.Thread] = threading.Thread(
+            name='grpc_local-thread', target=self.__poll_grpc)
+
+        self._sync_call_tp: Optional[concurrent.futures.Executor] = (
+            self._create_sync_call_tp(self._get_sync_tp_max_workers()))
+
```

Comment:
We should explore - moving this logic into lib_worker and not to put in proxy worker.